pkg/ccl/backupccl/backup_planning.go

```diff
@@ -399,6 +353,171 @@ func checkPrivilegesForBackup(
+
+		notice = fmt.Sprintf("%s BACKUP DATABASE, user %s will exclusively require the "+
+			"BACKUP privilege on database %s.",
+			preamble, p.User().Normalized(), strings.Join(dbsRequireBackupPrivilege, ", "))
+	} else if backupStmt.Targets.Tables.TablePatterns != nil {
+		// If a user is running `BACKUP TABLE` buffer all the tables that will require the `BACKUP` privilege.
+		tablesRequireBackupPrivilege := make([]string, 0)
+		for _, desc := range targetDescs {
+			switch desc := desc.(type) {
+			case catalog.TableDescriptor:
+				tablesRequireBackupPrivilege = append(tablesRequireBackupPrivilege, desc.GetName())
+			}
+		}
+
+		notice = fmt.Sprintf("%s BACKUP TABLE, user %s will exclusively require the "+
+			"BACKUP privilege on tables: %s.",
+			preamble,
+			p.User().Normalized(),
+			strings.Join(tablesRequireBackupPrivilege, ", "))
+	}
+
+	return notice
+}
+
+// checkPrivilegesForBackup is the driver method for privilege checks for all
+// flavours of backup. It checks that the user has sufficient privileges to read
+// the targets in the database, as well as use the External Storage URIs passed
+// in the backup statement.
+func checkPrivilegesForBackup(
+	ctx context.Context,
+	backupStmt *annotatedBackupStatement,
+	p sql.PlanHookState,
+	targetDescs []catalog.Descriptor,
+	to []string,
+) error {
+	// If the user is admin no further checks need to be performed.
+	hasAdmin, err := p.HasAdminRole(ctx)
+	if err != nil {
+		return err
+	}
+	if hasAdmin {
+		return nil
+	}
+
+	{
+		// Cluster and tenant backups require the `BACKUP` system privilege.
+		requiresBackupSystemPrivilege := backupStmt.Coverage() == tree.AllDescriptors ||
+			(backupStmt.Targets != nil && backupStmt.Targets.TenantID.IsSet())
+
+		var hasBackupSystemPrivilege bool
+		if p.ExecCfg().Settings.Version.IsActive(ctx, clusterversion.SystemPrivilegesTable) {
+			err := p.CheckPrivilegeForUser(ctx, syntheticprivilege.GlobalPrivilegeObject,
+				privilege.BACKUP, p.User())
+			hasBackupSystemPrivilege = err == nil
+		}
+
+		if requiresBackupSystemPrivilege && hasBackupSystemPrivilege {
+			return checkBackupDestinationPrivileges(ctx, p, to)
+		} else if requiresBackupSystemPrivilege && !hasBackupSystemPrivilege {
+			return errors.New("only users with admin role or the BACKUP " +
+				"system privilege are allowed to perform the backup")
+		}
+	}
+
+	// Target is only nil in cluster backups for which we have already checked
+	// privileges above.
+	if backupStmt.Targets == nil {
+		return errors.AssertionFailedf("programming error: target should be non-nil " +
+			"for database and table backups")
+	}
+
+	var hasRequiredBackupPrivileges bool
+	// If running a database backup, check that the user has the `BACKUP`
+	// privilege on all the target databases.
+	//
+	// TODO(adityamaru): In 23.1 a missing `BACKUP` privilege should return an
+	// error. In 22.2 we continue to check for old style privileges on the
+	// descriptors.
+	if backupStmt.Targets.Databases != nil {
```

Comment:
I like the if/else better, but now it is even more clear that these two branches are basically the same.

I would say it might be good to make lines 462-464 and lines 477-479 use the same structure.  I could also see just collapsing these into a single check.